### ISSUE TITLE
Better support for JSON arrays

### ISSUE THREAD

**Issue by Gerrrr** [confluentinc/ksql#9062]
**Is your feature request related to a problem? Please describe.**

A user in the [community Slack](https://confluentcommunity.slack.com/archives/C6UJNMY67/p1650902433356919) wanted to filter out/select items from a JSON array of objects.  Given records in the following format:

```sql
'[{"type": "AAA", "timestamp": "2021-09-27"}, {"type": "BBB", "timestamp": "2021-09-27"}, {"type": "AAA", "timestamp": "2021-09-27"}, {"type": "CCC", "timestamp": "2021-09-27"}]"}'
```

they wanted to select `timestamp` from the nested objects. In other words, they wanted to transform the input from above into

```sql
'[{"timestamp": "2021-09-27"}, {"timestamp": "2021-09-27"}, {"timestamp": "2021-09-27"}, {"timestamp": "2021-09-27"}]"}'
```

or

```sql
'["2021-09-27", "2021-09-27", "2021-09-27", "2021-09-27"]'
```

**Describe the solution you'd like**

To support that specific use case, we need to support 2 functions:
* a variation of `JSON_RECORDS` that converts a string with JSON array to an array of JSON strings - `JSON_ITEMS(String) -> Array<String>`
* [map](https://en.wikipedia.org/wiki/Map_(higher-order_function)) - a function that applies a function to each element of the given array - `MAP_F(ARRAY, lambda) -> ARRAY`

**UPD:** as @colinhicks pointed out, we already have map function that is called [TRANSFORM](https://docs.ksqldb.io/en/latest/developer-guide/ksqldb-reference/scalar-functions/#transform), so we only need `JSON_ITEMS` to satisfy the user requirements.

Then, a user could do something like 

```
MAP_F(JSON_ITEMS(message), (obj) => EXTRACTJSONFIELD(obj, '$.time`))
```

**Describe alternatives you've considered**

Ask a user to implement a UDF. This is a valid alternative that works. However, I think that the use case might be common enough to add those primitives.

**Comment by jnh5y** [comment]
Reminds me of https://github.com/confluentinc/ksql/issues/8798.  I wonder if it'd make sense to have a reduce for JSON arrays as well?  Or a flatMap?

**Comment by Gerrrr** [comment]
> I wonder if it'd make sense to have a reduce for JSON arrays as well?

TBH I would rather implement a function that transforms a string containing JSON array into a ksqlDB array of JSON strings. Then, we could reuse all existing functions working on arrays instead of implementing their counterparts that work only on JSON strings.

**Comment by jnh5y** [comment]
That makes sense.  Is there an example in Postgres (or other database)?  Either way, I like that idea.  Would you have time to open an issue for it?

**Comment by Gerrrr** [comment]
> Is there an example in Postgres (or other database)?

PostgreSQL's `json_array_elements` and `json_array_elements_text` seem very similar.

> 
That makes sense. Is there an example in Postgres (or other database)? Either way, I like that idea. Would you have time to open an issue for it?

I think this issue is good enough - it has a problem statement that explains why we need it and a spec for the function in the proposed solution.

**Comment by pallavnagar02** [comment]
Hi Team,

I am trying to flatten deeply nested complex data/events having STRUCT and ARRAY<STRUCT>, using EXPLODE for flattening nested Arrays but it gives  an error ...

SELECT
    ORDERID AS ORDER_ID,
    TIMESTAMP AS ORDER_TS,
    CUSTOMER->CUSTID AS CUST_ID,
    CUSTOMER->NAME AS CUST_NAME,
    CUSTOMER->CONTACT AS CUST_CONTACT,
    CUSTOMER->ADDRESS->CITY AS CUST_ADDR_CITY,
    CUSTOMER->ADDRESS->STATE AS CUST_ADDR_STATE,
    CUSTOMER->ADDRESS->ZIPCODE AS CUST_ADDR_ZIPCODE,
    EXPLODE(LINE_ITEMS)->SKU AS PROD_SKU,
    EXPLODE(LINE_ITEMS)->DESCRIPTION AS PROD_DETAILS,
    EXPLODE(LINE_ITEMS)->QTY AS PROD_QUANTITY,
    EXPLODE(LINE_ITEMS)->UNITPRICE AS PROD_UNIT_PRICE,
    PAYMENT->TYPE AS PAYMENT_TYPE,
    PAYMENT->LAST5 AS LAST_5_CARD_DIGITS,
    PAYMENT->EXPIRY AS CARD_EXPIRY
FROM
    ORDERS
EMIT CHANGES
LIMIT 5;

Error: Invalid Select: Can't find any functions with the name 'EXPLODE'. expression: EXPLODE(LINE_ITEMS)->SKU

but, when I select just the EXPLODE(ARRAY) data/records, it gives the desired output. 
SELECT 
  EXPLODE(LINE_ITEMS)->SKU AS PROD_SKU,
  EXPLODE(LINE_ITEMS)->DESCRIPTION AS PROD_DETAILS,
  EXPLODE(LINE_ITEMS)->QTY AS PROD_QUANTITY,
  EXPLODE(LINE_ITEMS)->UNITPRICE AS PROD_UNIT_PRICE
FROM 
	ORDERS 
EMIT CHANGES;

Please assist with this flattening of data.

**Comment by lucasprograms** [comment]
Thank you for adding this feature! Will it be released with 0.29?

**Comment by Gerrrr** [comment]
Hey @lucasprograms! Yeah, this feature will be included in 0.29.

**Comment by davidg2019** [comment]
Hi @Gerrrr, it seems that this feature is not available yet. When it will be released and in with version?

**Comment by Gerrrr** [comment]
Hey @davidg2019 ! 

`JSON_ITEMS` will be available in 0.29. The release work is in progress, but I can't share a specific date, unfortunately.
